pkg/assets/builder.go

```diff
@@ -127,32 +142,119 @@ func (a *AssetBuilder) RemapImage(image string) (string, error) {
 	}
 
 	fileAsset := &FileAsset{
-		File:              file,
-		CanonicalLocation: file,
+		FileURL: fileURL,
 	}
 
 	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
-		fileURL, err := url.Parse(file)
+		fileAsset.CanonicalFileURL = fileURL
+
+		normalizedFileURL, err := a.normalizeURL(fileURL)
 		if err != nil {
-			return "", fmt.Errorf("unable to parse file url %q: %v", file, err)
+			return nil, nil, fmt.Errorf("unable to parse file url %q: %v", normalizedFileURL, err)
 		}
 
-		fileRepo := strings.TrimSuffix(*a.AssetsLocation.FileRepository, "/")
-		fileAsset.File = fileRepo + fileURL.Path
+		fileAsset.FileURL = normalizedFileURL
+
+		glog.V(4).Infof("adding remapped file: %+v", fileAsset)
 	}
 
+	h, err := a.findHash(fileAsset)
+	if err != nil {
+		return nil, nil, err
+	}
+	fileAsset.SHAValue = h.Hex()
+
 	a.FileAssets = append(a.FileAssets, fileAsset)
+	glog.V(8).Infof("adding file: %+v", fileAsset)
+
+	return fileAsset.FileURL, h, nil
+}
+
+// TODO - remove this method as CNI does now have a SHA file
+
+// RemapFileAndSHAValue is used exclusively to remap the cni tarball, as the tarball does not have a sha file in object storage.
+func (a *AssetBuilder) RemapFileAndSHAValue(fileURL *url.URL, shaValue string) (*url.URL, error) {
+	if fileURL == nil {
+		return nil, fmt.Errorf("unable to remap an empty URL")
+	}
+
+	fileAsset := &FileAsset{
+		FileURL:  fileURL,
+		SHAValue: shaValue,
+	}
+
+	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
+		fileAsset.CanonicalFileURL = fileURL
+
+		normalizedFile, err := a.normalizeURL(fileURL)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse file url %q: %v", normalizedFile, err)
+		}
+
+		fileAsset.FileURL = normalizedFile
+		glog.V(4).Infof("adding remapped file: %q", fileAsset.FileURL.String())
+	}
+
+	a.FileAssets = append(a.FileAssets, fileAsset)
+
+	return fileAsset.FileURL, nil
+}
+
+// FindHash returns the hash value of a FileAsset.
+func (a *AssetBuilder) findHash(file *FileAsset) (*hashing.Hash, error) {
+	u := file.FileURL
+
+	// If the phase is "assets" we use the CanonicalFileURL,
+	// but during other phases we use the hash from the FileRepository or the base kops path.
+	// We do not want to just test for CanonicalFileURL as it is defined in
+	// other phases, but is not used to test for the SHA.
+	// This prevents a chicken and egg problem where the file is not yet in the FileRepository.
+	// Summary
+	// assets phase -> get the sha file from the source / CanonicalFileURL
+	// any other phase -> get the sha file from the kops base location or the FileRepository
+	if a.Phase == "assets" && file.CanonicalFileURL != nil {
+		u = file.CanonicalFileURL
+	}
+
+	if u == nil {
+		return nil, fmt.Errorf("file url is not defined")
+	}
+
+	// TODO This is old code that can be removed from the for loop.
+	// TODO This was in place in case we have different file extensions for sha files.
```

Comment:
Why have a loop if you do not need it? Why have code if you do not need it?

Reply:
It is a TODO comment.  I want to get the PR done, so let me know if you want me to remove this

Reply:
I think we have the loop because it's a clear way of parameterizing our code.  It's not a magic constant.  So I don't think we would remove the loop, and so I would remove the TODOs.  Or just add the ".sha1".  But the whole function is dirty in the diff anyway, so if you want TODO it, do it now :-)